gossipd/gossip_store.c

```diff
@@ -24,6 +31,15 @@ void gossip_store_append(struct gossip_store *gs, const u8 *msg)
 	u16 msglen = tal_len(msg);
 	beint16_t belen = CPU_TO_BE16(msglen);
 
+	/* FIXME: this method of detecting replayed messages is best effort
+	 * only. It should avoid doubling the store file size on every start,
+	 * but it'll allow the last few messages to be duplicated since replay
+	 * and write are async, and we'll think we are done replaying a bit too
+	 * early. */
+        /* Check if we are replaying the store */
+	if (lseek(gs->read_fd, 0, SEEK_CUR) < gs->replaysize)
+		return;
+
```

Comment:
Yech.  Definitely need to split this into a startup phase which reads all the gossip, and everything else.  In that phase the store won't write anything.

Reply:
I think the main concern is that we want a quick startup, so we just replay gossip when gossipd is not busy doing anything else, which is why `gossip_store` tries to allow reading and writing to interleave.

I think my concern about years-old gossip is still not addressed, too.

Reply:
That's exactly it. The startup phase would need to wait for txout resolution, i.e., the `io_loop` already needs to run, and we might be waiting for `bitcoind` to return blocks, which makes startup incredibly slow. I think we'll probably not get around the fact that we need to weave a `store` argument through the entire process to know whether this is an old or a new message.